docs/upgrade.md

```diff
@@ -3,5 +3,263 @@ title: Upgrade guide
+    return {
+      message: attributes.message,
+      service: attributes.serviceName,
+      environment: attributes.environment,
+      awsRegion: attributes.awsRegion,
+      correlationIds: {
+        awsRequestId: attributes.lambdaContext?.awsRequestId,
+        xRayTraceId: attributes.xRayTraceId,
+      },
+      lambdaFunction: {
+        name: attributes.lambdaContext?.functionName,
+        arn: attributes.lambdaContext?.invokedFunctionArn,
+        memoryLimitInMB: attributes.lambdaContext?.memoryLimitInMB,
+        version: attributes.lambdaContext?.functionVersion,
+        coldStart: attributes.lambdaContext?.coldStart,
+      },
+      logLevel: attributes.logLevel,
+      timestamp: this.formatTimestamp(attributes.timestamp),
+      logger: {
+        sampleRateValue: attributes.sampleRateValue,
+      },
+    };
+  }
+}
+
+export { MyCompanyLogFormatter };
+```
+
+In v2, the `formatAttributes` method is instead called with two arguments `attributes` and `additionalLogAttributes`. The `attributes` argument is the same as in v1, but the `additionalLogAttributes` argument is a plain object containing any [additional attributes you might have added](./core/logger.md#appending-persistent-additional-log-keys-and-values) to your logger. The method returns a `LogItem` object that contains all the attributes you want to include in the log output.
+
+```typescript hl_lines="1-2 5-8"
+import { LogFormatter, LogItem } from '@aws-lambda-powertools/logger';
+import type { LogAttributes, UnformattedAttributes } from '@aws-lambda-powertools/logger/types';
+
+class MyCompanyLogFormatter extends LogFormatter {
+  public formatAttributes(
+    attributes: UnformattedAttributes,
+    additionalLogAttributes: LogAttributes
+  ): LogItem {
+    const baseAttributes = {
+        message: attributes.message,
+        service: attributes.serviceName,
+        environment: attributes.environment,
+        awsRegion: attributes.awsRegion,
+        correlationIds: {
+            awsRequestId: attributes.lambdaContext?.awsRequestId,
+            xRayTraceId: attributes.xRayTraceId,
+        },
+        lambdaFunction: {
+            name: attributes.lambdaContext?.functionName,
+            arn: attributes.lambdaContext?.invokedFunctionArn,
+            memoryLimitInMB: attributes.lambdaContext?.memoryLimitInMB,
+            version: attributes.lambdaContext?.functionVersion,
+            coldStart: attributes.lambdaContext?.coldStart,
+        },
+        logLevel: attributes.logLevel,
+        timestamp: this.formatTimestamp(attributes.timestamp),
+        logger: {
+            sampleRateValue: attributes.sampleRateValue,
+        },
+    };
+    // Create a new LogItem with the base attributes
+    const logItem = new LogItem({ attributes: baseAttributes });
+    // Merge additional attributes
+    logItem.addAttributes(additionalLogAttributes);
+
+    return logItem;
+  }
+}
+
+export { MyCompanyLogFormatter };
+```
```

Comment:
In v1, `Logger` combined both [standard]((./core/logger.md#standard-structured-keys)) and [custom keys](./core/logger.md#appending-persistent-additional-log-keys-and-values) as a single argument, _e.g., `formatAttributes(attributes: UnformattedAttributes)`_. It expected a plain object with keys and values you wanted in the final log output.

In v2, you have more control over **standard** (`attributes`) and **custom keys** (`additionalLogAttributes`) in the `formatAttributes` method. Also, you now return a `LogItem` object to increase type safety when defining the final log output.

=== "Before"

    ```typescript hl_lines="5 8"
    import { LogFormatter } from '@aws-lambda-powertools/logger';
    import {
      LogAttributes,
      UnformattedAttributes,
    } from '@aws-lambda-powertools/logger/lib/types';

    class MyCompanyLogFormatter extends LogFormatter {
      public formatAttributes(attributes: UnformattedAttributes): LogAttributes {
        return {
          message: attributes.message,
          service: attributes.serviceName,
          environment: attributes.environment,
          awsRegion: attributes.awsRegion,
          correlationIds: {
            awsRequestId: attributes.lambdaContext?.awsRequestId,
            xRayTraceId: attributes.xRayTraceId,
          },
          lambdaFunction: {
            name: attributes.lambdaContext?.functionName,
            arn: attributes.lambdaContext?.invokedFunctionArn,
            memoryLimitInMB: attributes.lambdaContext?.memoryLimitInMB,
            version: attributes.lambdaContext?.functionVersion,
            coldStart: attributes.lambdaContext?.coldStart,
          },
          logLevel: attributes.logLevel,
          timestamp: this.formatTimestamp(attributes.timestamp),
          logger: {
            sampleRateValue: attributes.sampleRateValue,
          },
        };
      }
    }

    export { MyCompanyLogFormatter };
    ```

=== "After"

    ```typescript hl_lines="1-2 5-8"
    import { LogFormatter, LogItem } from '@aws-lambda-powertools/logger';
    import type { LogAttributes, UnformattedAttributes } from '@aws-lambda-powertools/logger/types';

    class MyCompanyLogFormatter extends LogFormatter {
      public formatAttributes(
        attributes: UnformattedAttributes,
        additionalLogAttributes: LogAttributes  // (1)!
      ): LogItem {  // (2)!
        const baseAttributes = {
            message: attributes.message,
            service: attributes.serviceName,
            environment: attributes.environment,
            awsRegion: attributes.awsRegion,
            correlationIds: {
                awsRequestId: attributes.lambdaContext?.awsRequestId,
                xRayTraceId: attributes.xRayTraceId,
            },
            lambdaFunction: {
                name: attributes.lambdaContext?.functionName,
                arn: attributes.lambdaContext?.invokedFunctionArn,
                memoryLimitInMB: attributes.lambdaContext?.memoryLimitInMB,
                version: attributes.lambdaContext?.functionVersion,
                coldStart: attributes.lambdaContext?.coldStart,
            },
            logLevel: attributes.logLevel,
            timestamp: this.formatTimestamp(attributes.timestamp),
            logger: {
                sampleRateValue: attributes.sampleRateValue,
            },
        };

        // Create a new LogItem with the base attributes
        const logItem = new LogItem({ attributes: baseAttributes });

        // Merge additional attributes
        logItem.addAttributes(additionalLogAttributes); // (3)!

        return logItem;
      }
    }

    export { MyCompanyLogFormatter };
    ```

    1. This new argument contains all [your custom keys](./core/logger.md#appending-persistent-additional-log-keys-and-values).
    2. `LogItem` is the new return object instead of a plain object.
    3. If you prefer adding at the initialization, use: <br/><br/> **`LogItem({persistentAttributes: additionalLogAttributes, attributes: baseAttributes})`**

tips: 

1. use `before/after` to group code changes like these, allowing you to bring upfront v1 vs v2 in text form. 
2. use code annotation to provide additional context like an alternative way to add both base + additional attributes.